### DISCUSSION TITLE
Artificially limit memory usage

### DISCUSSION THREAD

**Discussion by RealOrangeOne** [libvips/libvips#4106] [Help wanted]
I'm working in a somewhat memory constrained environment. Resizing large images (~92MP) consumes more memory than we'd like.

```
/usr/bin/time -f %M:%e python3 test.py
267912:1.25
```

Using `systemd-run`, I can limit the memory, and it works fine:

```
systemd-run --scope -p MemoryMax=10M --user /usr/bin/time -f %M:%e python3 test.py
37792:6.85
```

In my case, using less memory is better than being fast, so this works fine. However, this use case won't work for us in production, as we can't use `systemd` to limit the usage (in our case, from inside a Docker container)

Is there a way to artificially limit `libvips`'s memory usage? In a way which wouldn't cause it to crash on these larger images, just process slower?

**Comment by jcupitt** [comment]
Hi @RealOrangeOne,

Can you share your `test.py` benchmark? There are often things to tweak in that.

Limiting the threadpool can help, setting the env var `VIPS_CONCURRENCY=1` is the easiest way to do this.

libvips has an operation and image cache, you can disable this with `pyvips.cache_set_max(0)` near the start of your code.

The exact input and output image formats make a big difference. Interlaced JPG and PNG can be *very* expensive to process.

**Reply by RealOrangeOne** [reply]
Of course. It's intentionally simple, with the idea that once I get the basics working, I can deal with further functionality slowly on-top.

```python
import sys
import pyvips

pyvips.cache_set_max(0)

img = pyvips.Image.new_from_file(sys.argv[1], access=pyvips.Access.SEQUENTIAL)
img = img.thumbnail_image(500, height=500, size="both")

img.write_to_file("vips.png")
```

I'd already found some other threads which recommended `cache_set_max`, and whilst it did make a difference, it wasn't a sizeable one.

The test image I used is [this](https://github.com/user-attachments/assets/74aa1ce2-b97d-451d-8224-063d7340f18e) (I've checked GitHub's upload doesn't affect the result much)

`VIPS_CONCURRENCY` does limit the usage, and in this case drops it by about half, however that's still much higher than I'd like.

**Reply by jcupitt** [reply]
Don't use `thumbnail_image()` --- it's only there for emergencies. It prevents a lot of optimisations. With your test image:

```
$ /usr/bin/time -f %M:%e vips thumbnail_image big.png x.png 500 
277372:1.80
$ /usr/bin/time -f %M:%e vips thumbnail big.png x.png 500 
190776:0.84
$ VIPS_CONCURRENCY=1 /usr/bin/time -f %M:%e vips thumbnail big.png x.png 500 
93492:0.79
```

It makes a bigger difference with JPG images (`st-francis.jpg` is 30k x 30k pixels), if you plan to work on any of them, and even more for things like PDF, SVG and SVS.

```
$ /usr/bin/time -f %M:%e vips thumbnail_image st-francis.jpg x.jpg 500 
1400792:12.25
$ /usr/bin/time -f %M:%e vips thumbnail st-francis.jpg x.jpg 500 
318148:2.61
$ VIPS_CONCURRENCY=1 /usr/bin/time -f %M:%e vips thumbnail st-francis.jpg x.jpg 500 
275908:2.57
```

Are these images coming from a network? You can thumbnail directly from the pipe and not use the filesystem, it brings memory use down a bit further.

`VIPS_LEAK` is handy for memory testing:

```
$ VIPS_LEAK=1 VIPS_CONCURRENCY=1 /usr/bin/time -f %M:%e vips thumbnail big.png x.png 500 
vips_threadset_free: peak of 3 threads
memory: high-water mark 60.04 MB
93236:0.80
```

So the process is using 93mb of memory in total, libvips is allocating a peak of 60mb in pixel buffers, so perhaps 33mb is the libvips library.

`VIPS_INFO` and `VIPS_PROGRESS` can be useful too.

**Reply by RealOrangeOne** [reply]
Sometimes from the network, sometimes from disk. Ideally, I'd be able to use either.

If `thumbnail_image` isn't the right option, what's a better one? I've heard `.resize` is generally better, but occasionally I'll need to do more than just a simple scale (eg crop from center).

93888kb is much closer to what I'm after, so it's good to know vips _can_ do it, I just need to learn how to use it properly!

**Reply by jcupitt** [reply]
The libvips 8.16 docs have a new chapter listing some tips for devs:

https://github.com/libvips/libvips/blob/master/doc/Developer-checklist.md

Some of that might be useful too.

**Reply by jcupitt** [reply]
You just use plain `thumbnail("xxx.png", 500)`. It combines open and thumbnail in one operation, so the open can be optimised.

**Reply by RealOrangeOne** [reply]
Unfortunately, I already tried that:

```python
img = pyvips.Image.thumbnail(sys.argv[1], 500, height=500, size="both")
img.write_to_file("vips_efficient.png")
```

Which when run yields `167320:0.96` for the same test image. Still an improvement, but not to the level you got from the CLI, which feels strange.

**Reply by jcupitt** [reply]
For a network source, use `thumbnail_source`. There's some code in the pyvips examples directory:

https://github.com/libvips/pyvips/blob/master/examples/connection.py

https://github.com/libvips/pyvips/blob/master/examples/stream.py

**Reply by jcupitt** [reply]
Perhaps you have some other modules loaded? I tried:

```python
#!/usr/bin/env python3

import sys
import pyvips

pyvips.cache_set_max(0)
thumb = pyvips.Image.thumbnail(sys.argv[1], 500)
thumb.write_to_file(sys.argv[2])
```

Then with your test image:

```
$ VIPS_CONCURRENCY=1 /usr/bin/time -f %M:%e ./thumb.py ~/pics/big.png x.png
108848:0.93
```

So 33mb for libvips, 60mb for pixel buffers, 15mb for python.

**Reply by RealOrangeOne** [reply]
I think my previous test I'd forgotten to set `VIPS_CONCURRENCY` - I can indeed get your results from that script, which is great!

Is there any way to optimise the ~100MB down any further? Or is it just a fact of life for optimising images this large? I was hoping for something in the ~10MB range, which is clearly possible when explicitly limited.

**Reply by jcupitt** [reply]
python itself is 15mb, so you're not going to get lower than that heh.

You can remove a bit of buffering by dropping the tile height:

```
$ VIPS_LEAK=1 VIPS_CONCURRENCY=1 /usr/bin/time -f %M:%e vips thumbnail big.png x.png 500 --vips-tile-height=1
vips_threadset_free: peak of 3 threads
memory: high-water mark 16.81 MB
48632:0.76
```

So libvips is only allocating 17mb now. But this is just a thing for testing, it's not exposed in the API.

**Reply by RealOrangeOne** [reply]
I think I'll stick to the stable APIs ;) 100MB is still a lot better than what I have currently from Pillow, so I'll keep looking through the docs and probably accept that this is just a constraint to deal with.

**Reply by jcupitt** [reply]
Your test image is 275mb, so it's a big saving already.

**Comment by lovell** [comment]
> in our case, from inside a Docker container

Docker supports control groups - please see https://docs.docker.com/engine/containers/resource_constraints/

**Reply by RealOrangeOne** [reply]
It does, but only if you have full access to what's running the container, which we don't.
